If you have time, please add 1.21.4 support

As the title said, please add 1.21.4 support.
The plugin does seem to load but the few commands are not available and F3 + N shows no permissions.

Can confirm, currently does not work in version 1.21.4, using Paper.
`[20:12:47] [Server thread/WARN]: [F3NPerm] Task #39 for F3NPerm v3.5 generated an exception
nexus.slime.f3nperm.provider.ProviderException: Could not send packet!
    at F3NPerm-3.5.jar/nexus.slime.f3nperm.provider.ReflectionProvider_1_20_5.sendPacket(ReflectionProvider_1_20_5.java:46) ~[F3NPerm-3.5.jar:?]
    at F3NPerm-3.5.jar/nexus.slime.f3nperm.provider.NettyProvider.update(NettyProvider.java:36) ~[F3NPerm-3.5.jar:?]
    at F3NPerm-3.5.jar/nexus.slime.f3nperm.F3NPermListener.lambda$onJoin$0(F3NPermListener.java:24) ~[F3NPerm-3.5.jar:?]
    at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:78) ~[paper-1.21.4.jar:1.21.4-15-bb76819]
    at org.bukkit.craftbukkit.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:474) ~[paper-1.21.4.jar:1.21.4-15-bb76819]
    at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1805) ~[paper-1.21.4.jar:1.21.4-15-bb76819]
    at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1666) ~[paper-1.21.4.jar:1.21.4-15-bb76819]
    at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1331) ~[paper-1.21.4.jar:1.21.4-15-bb76819]
    at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:341) ~[paper-1.21.4.jar:1.21.4-15-bb76819]
    at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: nexus.slime.f3nperm.reflection.ReflectionException: Could not access field e on java.lang.Number
    at F3NPerm-3.5.jar/nexus.slime.f3nperm.reflection.Reflections.getPrivate(Reflections.java:203) ~[F3NPerm-3.5.jar:?]
    at F3NPerm-3.5.jar/nexus.slime.f3nperm.provider.ReflectionProvider_1_20_5.getNetworkManager(ReflectionProvider_1_20_5.java:26) ~[F3NPerm-3.5.jar:?]
    at F3NPerm-3.5.jar/nexus.slime.f3nperm.provider.ReflectionProvider_1_20_5.sendPacket(ReflectionProvider_1_20_5.java:41) ~[F3NPerm-3.5.jar:?]
    ... 9 more
Caused by: java.lang.NoSuchFieldException: e
    at java.base/java.lang.Class.getDeclaredField(Class.java:2782) ~[?:?]
    at io.papermc.reflectionrewriter.runtime.AbstractDefaultRulesReflectionProxy.getDeclaredField(AbstractDefaultRulesReflectionProxy.java:90) ~[reflection-rewriter-runtime-0.0.3.jar:?]
    at io.papermc.paper.pluginremap.reflect.PaperReflectionHolder.getDeclaredField(Unknown Source) ~[paper-1.21.4.jar:1.21.4-15-bb76819]
    at F3NPerm-3.5.jar/nexus.slime.f3nperm.reflection.Reflections.getPrivate(Reflections.java:199) ~[F3NPerm-3.5.jar:?]
    at F3NPerm-3.5.jar/nexus.slime.f3nperm.provider.ReflectionProvider_1_20_5.getNetworkManager(ReflectionProvider_1_20_5.java:26) ~[F3NPerm-3.5.jar:?]
    at F3NPerm-3.5.jar/nexus.slime.f3nperm.provider.ReflectionProvider_1_20_5.sendPacket(ReflectionProvider_1_20_5.java:41) ~[F3NPerm-3.5.jar:?]
    ... 9 more`

--- a/.gitignore
+++ b/.gitignore
@@ -3,3 +3,6 @@
 /build/
 /.idea/
 *.iml
+
+# Exclude run-paper server directory
+/run

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,10 +3,11 @@ import org.apache.tools.ant.filters.ReplaceTokens
 plugins {
     java
     idea
+    id("xyz.jpenilla.run-paper") version "2.3.1"
 }
 
 group = "nexus.slime"
-version = "3.5"
+version = "3.6"
 
 val targetJavaVersion = 8
 
@@ -70,4 +71,10 @@ tasks.processResources {
     filesMatching("config.yml") {
         filter(ReplaceTokens::class, mapOf("tokens" to props))
     }
+}
+
+// Add the option to run a server
+
+tasks.runServer {
+    minecraftVersion("1.21.5")
 }

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.1.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.14.1-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists

--- a/src/main/java/nexus/slime/f3nperm/BukkitVersion.java
+++ b/src/main/java/nexus/slime/f3nperm/BukkitVersion.java
@@ -49,7 +49,11 @@ public enum BukkitVersion {
     V1_20_4(43, "1.20.4-R0.1-SNAPSHOT"),
     V1_20_5(44, "1.20.5-R0.1-SNAPSHOT"),
     V1_20_6(45, "1.20.6-R0.1-SNAPSHOT"),
-    V1_21(46, "1.21-R0.1-SNAPSHOT");
+    V1_21(46, "1.21-R0.1-SNAPSHOT"),
+    V1_21_1(47, "1.21.1-R0.1-SNAPSHOT"),
+    V1_21_2(48, "1.21.2-R0.1-SNAPSHOT"),
+    V1_21_3(49, "1.21.3-R0.1-SNAPSHOT"),
+    V1_21_4(50, "1.21.4-R0.1-SNAPSHOT");
 
     public static BukkitVersion fromBukkitVersion() {
         String bukkitVersion = Bukkit.getServer().getBukkitVersion();

--- a/src/main/java/nexus/slime/f3nperm/F3NPermPlugin.java
+++ b/src/main/java/nexus/slime/f3nperm/F3NPermPlugin.java
@@ -119,7 +119,11 @@ public final class F3NPermPlugin extends JavaPlugin implements Listener {
         }
 
         // Just use the latest provider when we can't find a suitable one;
-        if (bukkitVersion == null || bukkitVersion.isGreaterOrEqualTo(BukkitVersion.V1_20_5)) {
+        if (bukkitVersion == null || bukkitVersion.isGreaterOrEqualTo(BukkitVersion.V1_21_3)) {
+            return new ReflectionProvider_1_21_3();
+        }
+
+        if (bukkitVersion.isGreaterOrEqualTo(BukkitVersion.V1_20_5)) {
             return new ReflectionProvider_1_20_5();
         }
 

--- a/src/main/java/nexus/slime/f3nperm/provider/ReflectionProvider_1_21_3.java
+++ b/src/main/java/nexus/slime/f3nperm/provider/ReflectionProvider_1_21_3.java
@@ -1,0 +1,22 @@
+package nexus.slime.f3nperm.provider;
+
+import nexus.slime.f3nperm.reflection.ReflectionException;
+import nexus.slime.f3nperm.reflection.Reflections;
+
+public class ReflectionProvider_1_21_3 extends ReflectionProvider_1_20_5 {
+
+    @Override
+    public Object makeStatusPacket(Object entityPlayer, byte status) throws ReflectionException {
+        return Reflections.make("net.minecraft.network.protocol.game.ClientboundEntityEventPacket(net.minecraft.world.entity.Entity,byte)", entityPlayer, status);
+    }
+
+    @Override
+    public boolean isStatusPacket(Object packet) throws ReflectionException {
+        return Reflections.resolve("net.minecraft.network.protocol.game.ClientboundEntityEventPacket").isInstance(packet);
+    }
+
+    @Override
+    public Object getPlayerConnection(Object entityPlayer) throws ReflectionException {
+        return Reflections.get(entityPlayer, "f");
+    }
+}

--- a/src/main/java/nexus/slime/f3nperm/provider/ReflectionProvider_1_21_3.java
+++ b/src/main/java/nexus/slime/f3nperm/provider/ReflectionProvider_1_21_3.java
@@ -4,7 +4,6 @@ import nexus.slime.f3nperm.reflection.ReflectionException;
 import nexus.slime.f3nperm.reflection.Reflections;
 
 public class ReflectionProvider_1_21_3 extends ReflectionProvider_1_20_5 {
-
     @Override
     public Object makeStatusPacket(Object entityPlayer, byte status) throws ReflectionException {
         return Reflections.make("net.minecraft.network.protocol.game.ClientboundEntityEventPacket(net.minecraft.world.entity.Entity,byte)", entityPlayer, status);

